File: output/qwen_2.5_72b_instruct/avatar/direct_translation/python/java/atcoder_ABC150_E.java

import java.util.Arrays;
import java.util.Scanner;

public class atcoder_ABC150_E {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int mod = 1000000007;
        int n = scanner.nextInt();
        int[] c = new int[n];
        for (int i = 0; i < n; i++) {
            c[i] = scanner.nextInt();
        }
        Arrays.sort(c);
        for (int i = 0; i < n / 2; i++) {
            int temp = c[i];
            c[i] = c[n - i - 1];
            c[n - i - 1] = temp;
        }
        long b = power(2, 2 * n - 2, mod);
        long a = (2 * b) % mod;
        long ans = 0;
        for (int i = 0; i < n; i++) {
            ans += (c[i] * ((a + (i * b) % mod) % mod)) % mod;
            ans %= mod;
        }
        System.out.println(ans);
    }

    public static long power(long x, long y, long mod) {
        long result = 1;
        x = x % mod;
        while (y > 0) {
            if ((y & 1) == 1) {
                result = (result * x) % mod;
            }
            y = y >> 1;
            x = (x * x) % mod;
        }
        return result;
    }
}

File: output/deepseek_chat/avatar/algo_based_translation/java/python/successful_translation/atcoder_ABC150_E.py

import sys

def power(x, n, mod):
    if n == 0:
        return 1
    val = power(x, n // 2, mod)
    val = (val * val) % mod
    if n % 2 == 1:
        val = (val * x) % mod
    return val

def main():
    data = sys.stdin.read().split()
    n = int(data[0])
    c = list(map(int, data[1:1+n]))
    mod = 1000000007
    
    if n == 1:
        result = (c[0] * 2) % mod
        print(result)
        return
        
    c.sort()
    b = power(2, n, mod)
    a = power(2, n - 2, mod)
    ans = 0
    
    for i in range(2, n + 2):
        val = (a * i) % mod
        val = (val * c[n + 1 - i]) % mod
        ans = (ans + val) % mod
        
    ans = (ans * b) % mod
    print(ans)

if __name__ == "__main__":
    main()

